google/api_core/retry_async.py

```diff
@@ -224,11 +236,29 @@ async def retry_wrapped_func(*args, **kwargs):
                 target,
                 self._predicate,
                 sleep_generator,
-                self._timeout,
+                timeout=self._timeout,
+                on_error=on_error,
+            )
+
+        @functools.wraps(func)
+        def retry_wrapped_stream(*args, **kwargs):
+            """A wrapper that iterates over target stream with retry."""
+            target = functools.partial(func, *args, **kwargs)
+            sleep_generator = exponential_sleep_generator(
+                self._initial, self._maximum, multiplier=self._multiplier
+            )
+            return retry_target_stream(
+                target,
+                self._predicate,
+                sleep_generator,
+                timeout=self._timeout,
                 on_error=on_error,
             )
 
-        return retry_wrapped_func
+        if self._is_stream:
```

Comment:
we can't use a function pointer here, like you did in the sync case (`retry_func = retry_target_stream if self._is_stream else retry_target`), so that we only need one wrapped function instead of two that are almost identical?

NB: I notice that you're making changing to this PR, so I can't find the code that I quoted in the current version, though it is in the PR that I cloned locally

Reply:
I think the only reason is because the existing `retry_wrapped_func` was an async function, and I was trying to add functionality without changing too much of the existing code.

But I just made a change to make `retry_wrapped_func` into a sync function that returns the retry coroutine directly instead of awaiting it, and that makes things cleaner. So let me know if that  works

Reply:
I like the new structure